integration/simulation/network/dockerenclave.go

```diff
@@ -13,41 +29,124 @@ import (
-	ethereum_mock "github.com/obscuronet/obscuro-playground/integration/ethereummock"
+)
+
+const (
+	enclaveDockerImg  = "obscuro_enclave"
+	enclaveAddress    = ":11000"
+	enclaveDockerPort = "11000/tcp"
 )
 
 // creates Obscuro nodes with their own enclave servers that communicate with peers via sockets, wires them up, and populates the network objects
 type basicNetworkOfNodesWithDockerEnclave struct {
-	ethNodes       []*ethereum_mock.Node
 	obscuroClients []*obscuroclient.Client
+	gethNetwork    *gethnetwork.GethNetwork
+	wallets        []wallet.Wallet
+	contracts      []string
+	workerWallet   wallet.Wallet
+	ctx            context.Context
+	client         *client.Client
+	containerIDs   map[string]string
 }
 
-func NewBasicNetworkOfNodesWithDockerEnclave() Network {
-	return &basicNetworkOfNodesWithDockerEnclave{}
+func NewBasicNetworkOfNodesWithDockerEnclave(wallets []wallet.Wallet, workerWallet wallet.Wallet, contracts []string) Network {
+	return &basicNetworkOfNodesWithDockerEnclave{
+		wallets:      wallets,
+		contracts:    contracts,
+		workerWallet: workerWallet,
+	}
 }
 
 // Create initializes Obscuro nodes with their own Dockerised enclave servers that communicate with peers via sockets, wires them up, and populates the network objects
 // TODO - Use individual Docker containers for the Obscuro nodes and Ethereum nodes.
 func (n *basicNetworkOfNodesWithDockerEnclave) Create(params *params.SimParams, stats *stats.Stats) ([]ethclient.EthClient, []*obscuroclient.Client, []string) {
+	// We create a Docker client.
+	n.ctx = context.Background()
+	cli, err := client.NewClientWithOpts()
+	if err != nil {
+		panic(err)
+	}
+	n.client = cli
+	// We check the required Docker images are available.
+	if !dockerImagesAvailable(n.ctx, cli) {
+		// We don't cause the test to fail here, because we want users to be able to run all the tests in the repo
+		// without having to build the Docker images.
+		println(fmt.Sprintf("This test requires the `%s` Docker image to be built using `dockerfiles/enclave.Dockerfile`. Terminating.", enclaveDockerImg))
+		return nil, nil, nil
+	}
+
+	// make sure the geth network binaries exist
+	path, err := gethnetwork.EnsureBinariesExist(gethnetwork.LatestVersion)
+	if err != nil {
+		panic(err)
+	}
+
+	// get wallet addresses to prefund them
+	walletAddresses := make([]string, len(n.wallets))
+	for i, w := range n.wallets {
+		walletAddresses[i] = w.Address().String()
+	}
+
+	// kickoff the network with the prefunded wallet addresses
+	n.gethNetwork = gethnetwork.NewGethNetwork(
+		params.StartPort,
+		params.StartPort+DefaultWsPortOffset,
+		path,
+		params.NumberOfNodes,
+		int(params.AvgBlockDuration.Seconds()),
+		walletAddresses,
+	)
+
+	tmpHostConfig := config.HostConfig{
+		L1NodeHost:          Localhost,
+		L1NodeWebsocketPort: n.gethNetwork.WebSocketPorts[0],
+		L1ConnectionTimeout: DefaultL1ConnectionTimeout,
+	}
+	tmpEthClient, err := ethclient.NewEthClient(tmpHostConfig)
+	if err != nil {
+		panic(err)
```

Comment:
@tudor-malene We should make sure we shutting down any existing components before panicking here and below. For example, this will skip terminating the Geth network, I think.

Reply:
unless I'm missing something, "TearDown" is deferred, so it will be called even if it panics here.

Reply:
Teardown is only deferred after `Create` has run - see `simulation_tester.go`.

Reply:
umm, what if we move `defer netw.TearDown()` before calling Create?

Reply:
Yes, that's definitely a good idea.